// ngx-input-file/esm2015/src/lib/input-file.module.js
import * as tslib_1 from "tslib";
var InputFileModule_1;
import { CommonModule } from '@angular/common';
import { DropZoneDirective } from './directives/drop-zone/drop-zone.directive';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { InputFileComponent } from './components/input-file/input-file.component';
import { InputFileService } from './services/input-file.service';
import { MatButtonModule } from '@angular/material/button';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { NgModule } from '@angular/core';
let InputFileModule = InputFileModule_1 = class InputFileModule {
    static forRoot(config) {
        return {
            ngModule: InputFileModule_1,
            providers: [
                InputFileService,
                { provide: 'config', useValue: config }
            ]
        };
    }
};
InputFileModule = InputFileModule_1 = tslib_1.__decorate([
    NgModule({
        declarations: [
            DropZoneDirective,
            InputFileComponent
        ],
        imports: [
            CommonModule,
            FormsModule,
            MatButtonModule,
            MatFormFieldModule,
            MatIconModule,
            MatInputModule,
            ReactiveFormsModule
        ],
        exports: [
            InputFileComponent
        ],
        providers: [
            InputFileService
        ],
        entryComponents: [InputFileComponent]
    })
], InputFileModule);
export { InputFileModule };
