// mat-video/esm5/lib/ui/mat-frame-by-frame-control/mat-frame-by-frame-control.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from "@angular/core";
var MatFrameByFrameControlComponent = /** @class */ (function () {
    function MatFrameByFrameControlComponent() {
        this.fps = 29.97;
    }
    /**
     * @return {?}
     */
    MatFrameByFrameControlComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} nbFrames
     * @return {?}
     */
    MatFrameByFrameControlComponent.prototype.seekFrames = /**
     * @param {?} nbFrames
     * @return {?}
     */
    function (nbFrames) {
        if (!this.video.paused) {
            this.video.pause();
        }
        /** @type {?} */
        var currentFrames = this.video.currentTime * this.fps;
        /** @type {?} */
        var newPos = (currentFrames + nbFrames) / this.fps + 0.00001;
        this.video.currentTime = newPos;
    };
    MatFrameByFrameControlComponent.decorators = [
        { type: Component, args: [{
                    selector: "mat-frame-by-frame-control",
                    template: "<button mat-icon-button (click)=\"seekFrames(-5)\">\r\n  <mat-icon>skip_previous</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(-1)\">\r\n  <mat-icon>arrow_left</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(1)\">\r\n  <mat-icon>arrow_right</mat-icon>\r\n</button>\r\n\r\n<button mat-icon-button (click)=\"seekFrames(5)\">\r\n  <mat-icon>skip_next</mat-icon>\r\n</button>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    MatFrameByFrameControlComponent.ctorParameters = function () { return []; };
    MatFrameByFrameControlComponent.propDecorators = {
        video: [{ type: Input }],
        fps: [{ type: Input }]
    };
    return MatFrameByFrameControlComponent;
}());
export { MatFrameByFrameControlComponent };
if (false) {
    /** @type {?} */
    MatFrameByFrameControlComponent.prototype.video;
    /** @type {?} */
    MatFrameByFrameControlComponent.prototype.fps;
}
