// ngx-input-file/esm5/src/lib/services/input-file.service.js
import * as tslib_1 from "tslib";
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var InputFileService = /** @class */ (function () {
    function InputFileService(_config) {
        this._config = _config;
    }
    Object.defineProperty(InputFileService.prototype, "config", {
        get: function () {
            return this._config;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Whether the limit is not reached.
     * @param files
     */
    InputFileService.prototype.limitGuard = function (files, fileLimit) {
        return files.length < fileLimit;
    };
    /**
     * Whether the file size is not bigger than the limit.
     * @param file
     * @param sizeLimit
     */
    InputFileService.prototype.sizeGuard = function (file, sizeLimit) {
        return !sizeLimit || file.size < sizeLimit * 1024 * 1024; // TODO : improve
    };
    /**
     * Whether the type of the file is enabled.
     * @param file
     * @param fileAccept
     */
    InputFileService.prototype.typeGuard = function (file, fileAccept) {
        var e_1, _a;
        var enabled = fileAccept == null;
        if (fileAccept) {
            var accept = fileAccept.replace('*', '');
            var types = accept.split(',');
            try {
                for (var types_1 = tslib_1.__values(types), types_1_1 = types_1.next(); !types_1_1.done; types_1_1 = types_1.next()) {
                    var type = types_1_1.value;
                    if (file.type.startsWith(type) || (type.charAt(0) === '.' && file.name != null && file.name.endsWith(type))) {
                        enabled = true;
                        break;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (types_1_1 && !types_1_1.done && (_a = types_1.return)) _a.call(types_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return enabled;
    };
    InputFileService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function InputFileService_Factory() { return new InputFileService(i0.ɵɵinject("config")); }, token: InputFileService, providedIn: "root" });
    InputFileService = tslib_1.__decorate([
        Injectable({
            providedIn: 'root'
        }),
        tslib_1.__param(0, Inject('config')),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], InputFileService);
    return InputFileService;
}());
export { InputFileService };
