// mat-video/esm2015/lib/directives/mat-video-source.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Host, Input } from "@angular/core";
import { MatVideoComponent } from "../video.component";
export class MatVideoSourceDirective {
    /**
     * @param {?} matVideoComponent
     * @param {?} el
     */
    constructor(matVideoComponent, el) {
        this.matVideoComponent = matVideoComponent;
        this.el = el;
        this.src = null;
        this.type = null;
        this.init = true;
        this.video = matVideoComponent;
        this.source = el.nativeElement;
        this.init = false;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.source.src = this.src;
        this.source.type = this.type;
        if (!this.init) {
            this.video.load();
        }
    }
}
MatVideoSourceDirective.decorators = [
    { type: Directive, args: [{
                selector: "[matVideoSource]"
            },] }
];
/** @nocollapse */
MatVideoSourceDirective.ctorParameters = () => [
    { type: MatVideoComponent, decorators: [{ type: Host }] },
    { type: ElementRef }
];
MatVideoSourceDirective.propDecorators = {
    src: [{ type: Input }],
    type: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MatVideoSourceDirective.prototype.src;
    /** @type {?} */
    MatVideoSourceDirective.prototype.type;
    /**
     * @type {?}
     * @private
     */
    MatVideoSourceDirective.prototype.init;
    /**
     * @type {?}
     * @private
     */
    MatVideoSourceDirective.prototype.video;
    /**
     * @type {?}
     * @private
     */
    MatVideoSourceDirective.prototype.source;
    /**
     * @type {?}
     * @private
     */
    MatVideoSourceDirective.prototype.matVideoComponent;
    /**
     * @type {?}
     * @private
     */
    MatVideoSourceDirective.prototype.el;
}
