// ngx-input-file/esm2015/src/lib/services/input-file.service.js
import * as tslib_1 from "tslib";
import { Inject, Injectable } from '@angular/core';
import * as i0 from "@angular/core";
let InputFileService = class InputFileService {
    constructor(_config) {
        this._config = _config;
    }
    get config() {
        return this._config;
    }
    /**
     * Whether the limit is not reached.
     * @param files
     */
    limitGuard(files, fileLimit) {
        return files.length < fileLimit;
    }
    /**
     * Whether the file size is not bigger than the limit.
     * @param file
     * @param sizeLimit
     */
    sizeGuard(file, sizeLimit) {
        return !sizeLimit || file.size < sizeLimit * 1024 * 1024; // TODO : improve
    }
    /**
     * Whether the type of the file is enabled.
     * @param file
     * @param fileAccept
     */
    typeGuard(file, fileAccept) {
        let enabled = fileAccept == null;
        if (fileAccept) {
            const accept = fileAccept.replace('*', '');
            const types = accept.split(',');
            for (const type of types) {
                if (file.type.startsWith(type) || (type.charAt(0) === '.' && file.name != null && file.name.endsWith(type))) {
                    enabled = true;
                    break;
                }
            }
        }
        return enabled;
    }
};
InputFileService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function InputFileService_Factory() { return new InputFileService(i0.ɵɵinject("config")); }, token: InputFileService, providedIn: "root" });
InputFileService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__param(0, Inject('config')),
    tslib_1.__metadata("design:paramtypes", [Object])
], InputFileService);
export { InputFileService };
